### PR TITLE
try commit 19/12

### PR DESCRIPTION
**Pull request by Watxesnavo**

### CODE REVIEW

**Review by aleksandr-bevza — CHANGES_REQUESTED** [review]
image
Looks like some tests fail. Fix this tests please

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -5,10 +5,27 @@
...
+            throw new CustomException("empty login");
+        }
+        if (user.getPassword() == null) {
+            throw new CustomException("empty password");
+        }
+        if (user.getAge() < MIN_AGE) {
+            throw new CustomException("the age is under 18");
+        }
+        if (user.getPassword().length() < MIN_PASS_LENGTH) {
+            throw new CustomException("password length must be more than 6");
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:24]: 1. Use "MIN_PASSWORD_LENGTH" in message
2. Message is not correct. You say: "password length must be more than 6" It means that length 6 is not correct, but length 7 is correct

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -5,10 +5,27 @@
...
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new CustomException("that user already exist in storage");
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:27]: Include user's login to this message

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static final int MIN_AGE = 18;
+    private static final int MIN_PASS_LENGTH = 6;
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:11]:
```suggestion
    private static final int MIN_PASSWORD_LENGTH = 6;
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        user.setPassword(password);
+    }
+
+    @Test
+    void register_UserExistsAlready_Ok() {
+        User user1 = registrationService.register(user);
+        Assertions.assertThrows(CustomException.class, () -> {
+            registrationService.register(user1);
+        });
+        Storage.people.remove(user1);
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:32]: Delete it in all places. Use @AfterEach instead
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    private final String password = "Sukhov28";
+
+    @BeforeEach
+    void setUp() {
+        user = new User();
+        user.setAge(age++);
+        user.setLogin(login);
+        user.setPassword(password);
+    }
+
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:25]:
```suggestion
@AfterEach
void tearDown() {
    Storage.people.clear();
}
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void register_UserExistsAlready_Ok() {
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:27]: Is it really "ok"? (last part of method's name). Looks like we check "notOk" behavior in this test

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void register_UserAge_Ok() {
+        user.setAge(19);
+        User user1 = registrationService.register(user);
+        Assertions.assertTrue(user1.getAge() > MIN_AGE);
+        Storage.people.remove(user1);
+    }
+
+    @Test
+    void register_UserAgeNotOk_ThrowsException_Ok() {
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:44]: image

Please, follow the naming requirements for this homework:
```suggestion
    void register_AgeLessThan18_NotOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        user.setAge(12);
+        Assertions.assertThrows(CustomException.class, () -> {
+            registrationService.register(user);
+        });
+        Storage.people.remove(user);
+    }
+
+    @Test
+    void register_UserLoginNull_ThrowsException_Ok() {
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:53]: Please, follow the naming requirements for this homework:
Fix in all places
```suggestion
    void register_LoginIsNull_NotOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            registrationService.register(user1);
+        });
+        Storage.people.remove(user1);
+    }
+
+    @Test
+    void register_UserAge_Ok() {
+        user.setAge(19);
+        User user1 = registrationService.register(user);
+        Assertions.assertTrue(user1.getAge() > MIN_AGE);
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:39]:
```suggestion
        Assertions.assertTrue(user1.getAge() >= MIN_AGE);
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    void register_UserLogin_NotNull() {
+        User user1 = registrationService.register(user);
+        Assertions.assertNotNull(user1.getLogin());
+        Storage.people.remove(user1);
+    }
+
+    @Test
+    void register_UserPasswordLength_Ok() {
+        User user1 = registrationService.register(user);
+        Assertions.assertTrue(user1.getPassword().length() > MIN_PASS_LENGTH);
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:71]:
```suggestion
        Assertions.assertTrue(user1.getPassword().length() >= MIN_PASS_LENGTH);
```

**Review by aleksandr-bevza — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationService.java [hunk 11]
```diff
@@ -3,5 +3,5 @@
 import core.basesyntax.model.User;
 
 public interface RegistrationService {
-    User register(User user);
+    User register(User user) throws RegisterException;
```
Comment (aleksandr-bevza) [inline src/main/java/core/basesyntax/service/RegistrationService.java:6]: No need to add unchecked exception to method signature (throws)
```suggestion
    User register(User user);
```

**Review by aleksandr-bevza — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 12]
```diff
@@ -22,14 +22,14 @@
...
             <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter-api</artifactId>
+            <artifactId>junit-jupiter-engine</artifactId>
             <version>${junit-jupiter.version}</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter-engine</artifactId>
-            <version>${junit-jupiter.version}</version>
```
Comment (nacenik) [inline pom.xml:32]: don't change pom file.

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 13]
```diff
@@ -5,10 +5,29 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_AGE = 18;
+    private static final int MIN_PASS_LENGTH = 6;
```
Comment (nacenik) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:9]: Make constant name more informative

**Review by nacenik — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Test
+    void register_UserPasswordLength_Ok() {
+        User user1 = registrationService.register(user);
+        Assertions.assertTrue(user1.getPassword().length() >= MIN_PASSWORD_LENGTH);
+    }
+
+    @Test
+    void register_UserPasswordLength_ThrowsException_Ok() {
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:76]:
```suggestion
    void register_SmallPasswordLength_NotOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        user.setPassword("123");
+        Assertions.assertThrows(RegisterException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_UserPasswordNull_ThrowsException_Ok() {
```
Comment (aleksandr-bevza) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:84]:
```suggestion
    void register_UserPasswordNull_NotOk() {
```

**Review by aleksandr-bevza — APPROVED** [review]

**Pull request closed without merging**